examples/common/convert_c4.py

```diff
@@ -4,105 +4,359 @@
+        return len(it.kwargs['filepaths'])  # type: ignore
+
+    def __iter__(self):
+        worker_info = get_worker_info()
+        if worker_info:
+            num_workers = worker_info.num_workers
+            worker_id = worker_info.id
+            it = self.dataset._ex_iterable  # type: ignore
+            shards = it.kwargs['filepaths']  # type: ignore
+            assert len(shards) % num_workers == 0
+            it.kwargs['filepaths'] = shards[  # type: ignore
+                worker_id::num_workers]
+        return iter(self.dataset)
+
+    def _build_loader(self) -> None:
+        # Multiple workers is only supported on linux machines
+        if 'linux' in platform.platform().lower():
+            num_workers = min(64, self.num_shards())  # type: ignore
+        else:
+            num_workers = 0
+
+        # If using multiple workers, configure each worker to prefetch as many samples as it can, up to
+        # the aggregate device batch size
+        # If not using workers, the torch DataLoader expects the default value for prefetch_factor,
+        # which non-intuitively must be 2.
+        prefetch_factor = max(1, 2 * self.batch_size //
+                              num_workers) if num_workers > 0 else 2
+
+        self.loader = DataLoader(  # type: ignore
+            dataset=self.dataset,
+            sampler=None,
+            batch_size=self.batch_size,
+            num_workers=num_workers,
+            prefetch_factor=prefetch_factor,
+        )
+
+    def generate_samples(self) -> Iterable[Dict[str, bytes]]:
+        """Generator over each dataset sample.
+
+        Args:
+            samples (IterableDataset): An iterable dataset that is multi-worker compatible
+
+        Yields:
+            Sample dicts.
+        """
+        total_chars = 0
+        n_samples = 0
+        for batch in self.loader:
+            keys = list(batch.keys())
+            current_bs = len(batch[keys[0]])
+            for idx in range(current_bs):
+                total_chars += len(batch['text'][idx].encode('utf-8'))
+                yield {
+                    key: batch_values[idx].encode('utf-8')
+                    for key, batch_values in batch.items()
+                }
+                n_samples += 1
+                if n_samples == self.expected_num_samples:
+                    print(f'{total_chars=}')
+                    return
+
+
+class ConcatC4(ShardedC4):
 
-    return args.parse_args()
+    def __init__(self, sep_text: str, *args, **kwargs):
+        self.sep_text = sep_text
+        super().__init__(*args, **kwargs)
 
+    def generate_samples(self) -> Iterable[Dict[str, bytes]]:
+        """Generator over each dataset sample.
 
-def build_hf_c4_dataset(split: str) -> IterableDataset:
+        Args:
+            samples (IterableDataset): An iterable dataset that is multi-worker compatible
+
+        Yields:
+            Sample dicts.
+        """
+
+        n_samples = 0
+        buffer = {}
```

Comment:
seems like this doesn't really need to be a dict and can just be a single string?